sphinxdocs/sphinx.bzl

```diff
@@ -0,0 +1,191 @@
+            in. Each format will become an output group.
+        strip_prefix: (str) A prefix to remove from the file paths of the
+            source files. e.g., given `//docs:foo.md`, stripping `docs/`
+            makes Sphinx see `foo.md` in its generated source directory.
+    """
+
+    _sphinx_docs(name = name, **kwargs)
+
+    html_name = "_{}_html".format(name)
+    native.filegroup(
+        name = html_name,
+        srcs = [name],
+        output_group = "html",
+    )
+    py_binary(
+        name = name + ".serve",
+        srcs = [_SPHINX_SERVE_MAIN_SRC],
+        main = _SPHINX_SERVE_MAIN_SRC,
+        data = [html_name],
+        args = [
+            "$(execpath {})".format(html_name),
+        ],
+    )
+
+def _sphinx_docs_impl(ctx):
+    source_dir_path, inputs = _create_sphinx_source_tree(ctx)
+    inputs.append(ctx.file.config)
+
+    outputs = {}
+    for format in ctx.attr.formats:
+        output_dir = _run_sphinx(
+            ctx = ctx,
+            format = format,
+            source_path = source_dir_path,
+            output_prefix = "{}/_build".format(ctx.label.name),
+            inputs = inputs,
+        )
+        outputs[format] = output_dir
+    return [
+        DefaultInfo(files = depset(outputs.values())),
+        OutputGroupInfo(**{
+            format: depset([output])
+            for format, output in outputs.items()
+        }),
+    ]
+
+_sphinx_docs = rule(
+    implementation = _sphinx_docs_impl,
+    attrs = {
+        "srcs": attr.label_list(
+            allow_files = True,
+            doc = "Doc source files for Sphinx.",
+        ),
+        "sphinx": attr.label(
+            executable = True,
+            cfg = "exec",
+            mandatory = True,
+            doc = "Sphinx binary to generate documentation.",
+        ),
+        "config": attr.label(
+            allow_single_file = True,
+            mandatory = True,
+            doc = "Config file for Sphinx",
+        ),
+        "formats": attr.string_list(doc = "Output formats for Sphinx to create."),
+        "strip_prefix": attr.string(doc = "Prefix to remove from input file paths."),
+    },
+)
+
+def _create_sphinx_source_tree(ctx):
+    # Sphinx only accepts a single directory to read its doc sources from.
+    # Because plain files and generated files are in different directories,
+    # we need to merge the two into a single directory.
+    source_prefix = ctx.label.name + "/_sources"
+    source_marker = ctx.actions.declare_file(source_prefix + "/__marker")
+    ctx.actions.write(source_marker, "")
+    sphinx_source_dir_path = paths.dirname(source_marker.path)
+    sphinx_source_files = []
+    for orig in ctx.files.srcs:
+        source_rel_path = orig.short_path
+        if source_rel_path.startswith(ctx.attr.strip_prefix):
+            source_rel_path = source_rel_path[len(ctx.attr.strip_prefix):]
+
+        sphinx_source = ctx.actions.declare_file(source_prefix + "/" + source_rel_path)
```

Comment:
nit: what about using `bazel-skylib` `paths.join` here and elsewhere in the PR?

Reply:
Done